### PR TITLE
nixos/zapret: extra features

### PR DESCRIPTION
**Pull request by voronind-com**
Added some of the requested by users features:
- [x] "Full" http processing mode.
	- [x] Tested.
- [x] Udp support.
	- [x] Tested.
- [x] ipv6 support.
	- [x] Tested ipv4 only.

Planned later:
* ipset support.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Nishimara — APPROVED** [review]
LGTM. Don't forget to make branch-off before making pull request next time!

**Comment by voronind-com** [comment]
I added examples to new options. It is ready for now. I'll try to implement the requested ipv6 support as soon as I get access to it.

**Review by TheMaxMur — APPROVED** [review]

**Comment by roman3pm** [comment]
I'm really interested in ipv6 feature. How can I help to test this? In case of implementation I think we can do something like this:
```nix
(lib.mkIf cfg.configureFirewall {
  networking.firewall.extraCommands =
    let
      httpParams = lib.optionalString (
        cfg.httpMode == "first"
      ) "-m connbytes --connbytes-dir=original --connbytes-mode=packets --connbytes 1:6";

      udpPorts = lib.concatStringsSep "," cfg.udpPorts;
      iptablesCommands = [ "iptables" ] ++ lib.optional (cfg.ipv6) "ip6tables";
      mkIptablesRules =
        command:
        ''
          ${command} -t mangle -I POSTROUTING -p tcp --dport 443 -m connbytes --connbytes-dir=original --connbytes-mode=packets --connbytes 1:6 -m mark ! --mark 0x40000000/0x40000000 -j NFQUEUE --queue-num ${qnum} --queue-bypass
        ''
        + lib.optionalString (cfg.httpSupport) ''
          ${command} -t mangle -I POSTROUTING -p tcp --dport 80 ${httpParams} -m mark ! --mark 0x40000000/0x40000000 -j NFQUEUE --queue-num ${qnum} --queue-bypass
        ''
        + lib.optionalString (cfg.udpSupport) ''
          ${command} -t mangle -A POSTROUTING -p udp -m multiport --dports ${udpPorts} -m mark ! --mark 0x40000000/0x40000000 -j NFQUEUE --queue-num ${qnum} --queue-bypass
        '';
    in
    lib.concatMapStrings mkIptablesRules iptablesCommands;
})
```

**Comment by voronind-com** [comment]
> I'm really interested in ipv6 feature. How can I help to test this? In case of implementation I think we can do something like this:
> 
> ```nix
> (lib.mkIf cfg.configureFirewall {
>   networking.firewall.extraCommands =
>     let
>       httpParams = lib.optionalString (
>         cfg.httpMode == "first"
>       ) "-m connbytes --connbytes-dir=original --connbytes-mode=packets --connbytes 1:6";
> 
>       udpPorts = lib.concatStringsSep "," cfg.udpPorts;
>       iptablesCommands = [ "iptables" ] ++ lib.optional (cfg.ipv6) "ip6tables";
>       mkIptablesRules =
>         command:
>         ''
>           ${command} -t mangle -I POSTROUTING -p tcp --dport 443 -m connbytes --connbytes-dir=original --connbytes-mode=packets --connbytes 1:6 -m mark ! --mark 0x40000000/0x40000000 -j NFQUEUE --queue-num ${qnum} --queue-bypass
>         ''
>         + lib.optionalString (cfg.httpSupport) ''
>           ${command} -t mangle -I POSTROUTING -p tcp --dport 80 ${httpParams} -m mark ! --mark 0x40000000/0x40000000 -j NFQUEUE --queue-num ${qnum} --queue-bypass
>         ''
>         + lib.optionalString (cfg.udpSupport) ''
>           ${command} -t mangle -A POSTROUTING -p udp -m multiport --dports ${udpPorts} -m mark ! --mark 0x40000000/0x40000000 -j NFQUEUE --queue-num ${qnum} --queue-bypass
>         '';
>     in
>     lib.concatMapStrings mkIptablesRules iptablesCommands;
> })
> ```

Is replacing `iptables` with `ip6tables` is the only requirement for zapret? If so, this should be easy. I'll try to draft it within a day, could you test it, please? I'll ping you on GH.

**Comment by Nishimara** [comment]
Try looking into `ip46tables`. Firewall have this hack to execute commands with both `iptables` and `ip6tables`.

**Comment by roman3pm** [comment]
> Is replacing `iptables` with `ip6tables` is the only requirement for zapret? If so, this should be easy. I'll try to draft it within a day, could you test it, please? I'll ping you on GH.
Not a replacement, but an addition of exactly the same rule using `ip6tables`. Sure I can test this, but do not know exactly how) Any guides are welcome here

**Comment by Nishimara** [comment]
@voronind-com
Maybe something like this?
```nix
let
  ...
  tables = if cfg.ipv6Support then "ip46tables" else "iptables";
in
  ...
  networking.firewall.extraCommands = "${tables} -t mangle -I POSTROUTING -p tcp --dport 443 -m connbytes --connbytes-dir=original --connbytes-mode=packets --connbytes 1:6 -m mark ! --mark 0x40000000/0x40000000 -j NFQUEUE --queue-num ${qnum} --queue-bypass";
```

**Comment by voronind-com** [comment]
Thanks. I'm uber tired today, I'll update this pr tomorrow morning. Also will help roman3pm with testing.

**Comment by voronind-com** [comment]
Sadly I have no ip46tables in my system, so I guess it is not included by default. Referencing another pkg also might not be a good idea. I'm gonna do additional rule with ip6tables now.

**Comment by voronind-com** [comment]
@roman3pm It's ready for testing. Actually super easy. All you need to do is add this to your nixos config:
```text
disabledModules = [ "services/networking/zapret.nix" ];
```
Then copy-paste https://github.com/NixOS/nixpkgs/blob/2cde9592dfba6ea3a95d6cece22a328927908864/nixos/modules/services/networking/zapret.nix and import it as a regular module. Switch and test.

**Comment by voronind-com** [comment]
Fixed a typo. Works fine with `ip46tables` on ipv4 network. Marking as ready.

**Review by SnakeOPM — APPROVED** [review]

**Review by SnakeOPM — APPROVED** [review]

File: nixos/modules/services/networking/zapret.nix [hunk 1]
```diff
@@ -96,34 +98,79 @@ in
...
           {
             assertion = (builtins.length cfg.params) != 0;
             message = "You have to specify zapret parameters. See the params option's description.";
           }
+          {
+            assertion = cfg.udpSupport -> (builtins.length cfg.udpPorts) != 0;
+            message = "You have to specify UDP ports or disable UDP support.";
+          }
+          {
+            assertions = !cfg.configureFirewall || !config.networking.nftables.enable;
```
[the remark this answers is not in the document]
Comment (voronind-com) [reply nixos/modules/services/networking/zapret.nix]: Yes, I fixed it in my previous message with "fixed a typo". Thanks for the spot!

[the remark this answers is not in the document]
Comment (voronind-com) [reply nixos/modules/services/networking/zapret.nix]: My bad, I see it now.

**Comment by ReLoneR1** [comment]
Hello @Nishimara, @voronind-com . I wanted to add --ipset option. I created https://github.com/NixOS/nixpkgs/pull/358168, could you please check it out

**Comment by voronind-com** [comment]
> Hello @Nishimara, @voronind-com . I wanted to add --ipset option. I created #358168, could you please review it

Hello! I see. I'll research that option a bit and sure will add it in this PR, okay? Or do you want to author on it?

**Comment by voronind-com** [comment]
> sure, add it to this pr, i don`t need to author on it. Did you think about how a QUIC parameters could be added?

Not yet. I still want to iteratively add small features.

**Comment by voronind-com** [comment]
As for `--ipset` option there's also`--ipset-exclude`. We already occupied the `whitelist` and `blacklist` option names. So we have a few options:

1. Theoretically we could detect IP addresses in them and move to a different list. Tho I have no idea how to implement that, but I could try.
2. Create `filterMode` option with values `domain` and `ip`. But this disallows using both. I'm unsure whether Zapret even supports using both at the same time.
3. Straightforward creation of a new pair: `ipwhitelist` and `ipblacklist` or something like that. This would theoretically allow the usage of both filters.
4. ???

@Nishimara what do you think?

**Comment by Nishimara** [comment]
I think it's better to move it into next PR, as soon as this merged. Otherwise we block well-tested features implemented here from going to unstable.

**Review by TheMaxMur — APPROVED** [review]

**Review by azahi — APPROVED** [review]

**Pull request merged**